source/extensions/quic_listeners/quiche/platform/quic_mem_slice_storage_impl.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include "common/buffer/buffer_impl.h"
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
+#include "quiche/quic/platform/api/quic_iovec.h"
+#include "quiche/quic/platform/api/quic_mem_slice_span.h"
+
+namespace quic {
+
+// QuicMemSliceStorageImpl wraps a MemSlice vector.
+class QuicMemSliceStorageImpl {
+public:
+  QuicMemSliceStorageImpl(const struct iovec* iov, int iov_count, QuicBufferAllocator* allocator,
+                          const QuicByteCount max_slice_len);
+
+  QuicMemSliceStorageImpl(const QuicMemSliceStorageImpl& other)
+      : buffer_(static_cast<const Envoy::Buffer::Instance&>(other.buffer_)) {}
```

Comment:
cast does not look necessary.

Reply:
I don't understand your response. I don't think the cast is needed because other.buffer_ has the same type as buffer_.

Reply:
W/o static_cast this would be a copy construction. But OwnedImpl doesn't have copy constructor, instead, it has `OwnedImpl(const Instance& data)`.  
Since you already patch this PR, maybe you can try removing static_cast in your local to see the compile error message, if this is not clear to you.

Reply:
I see. The cast looks wrong then. How can you cast an OwnedImpl to an Instance?

Also whenever I put anything into the body, it fails to compile with an alignment error.

Also, if I delete that constructor altogether, the tests compile and pass. Maybe just do that?

Reply:
Never mind; it's casting to its super-super class.  I need to now think about how this interacts with the two alternate implementations of Buffer.

Reply:
and it seems pretty weird to double-upcast. Can you just instead add a copy-ctor to OwnedImpl? And what happens when Envoy is run using new buffers?

Reply:
I think copy constructor was left out intentionally. The slices in OwnedImpl has unique ownership. 
`OwnedImpl(const Instance& data)` just copies over the data, while a copy constructor need to deal with ownership.

Reply:
I don't understand the point about ownership, but this is just style and not having unexpected looking code, so meh. I think given this state I would prefer adding the copy-ctor to the subclass.
 
But the larger issue is with the two alternate buffer strategies. It looks like we are probably going to switch to the new buffer implementation, and you are directly referencing the old one. And it is command-line switchable. I think we don't want to mix them in a single process, so by you hard-coding the reference to the old buffer implementation I think that will caused mixed usage.

Reply:
I added copy constructor to OwnedImpl which only copies over data but leaves out use_old_impl_.

As to the two alternative implementation. Are you referring to evbuffer and SliceDeque? They are hidden behind the OwnedImpl. I think whether OwnedImpl uses one or the other as underlying container doesn't matter. Or are you saying there is an alternative to OwnedImpl?

Reply:
tag @mattklein123 for the reason why I added copy constructor and assignment to OwnedImpl.